internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,142 @@ import (
-	Charge(cost types.GasUnits) error
-	SampleChainRandomness(sampleHeight *types.BlockHeight) ([]byte, error)
+// Runtime2 is the VM.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: for @go-filecoin, should we turn the offset into a RandomnessEvent?
+	Randomness(epoch Epoch, offset uint64) Randomness
+	Send(input InvocInput) result.InvocOutput
+	// SharedStorage is an IPLD store shared by all actors.
+	// Review: for @go-filecoin, we can skip this until we have actor code that actually needs this
+	// Review: for @spec, Add expiration (Epoch's) on PUT? (the question is mostly for global, but the local store might want to advise on this too).
+	// Review: for @spec, how will gas accounting work on global? PUT fee, GET fee, OWNER Fee?
+	SharedStorage() IpldStorage2
+}
+
+// DispatchContext is the context the actor dispatcher receives.
+// Review: for @go-filecoin, do we want to have the `ctx.ValidateCaller(blahblah)` at the top of each method, or hidden away on the dispatch table?
+type DispatchContext interface {
+	// Review: for @go-filecoin, simplified the API to be just one method.
+	ValidateCaller(CallerPattern)
+	// InvocationContext is the context to be passed down the actor method.
+	InvocationContext() InvocationContext
+}
+
+// InvocationContext is passed to the actors on each method call..
+type InvocationContext interface {
+	// Review: for @go-filecoin, composition or inheritance?
+	Runtime() Runtime2
+	Caller() address.Address
+	StateHandler() ActorStateHandler
+
+	// Review: for @go-filecoin+@spec, renamed `ValueSupplied` to `ValueReceived`. The "supplied" is not clear by whom.
+	ValueReceived() types.AttoFIL
+	Balance() types.AttoFIL
+
+	LocalStorage() IpldStorage2
+}
 
-	CreateNewActor(addr address.Address, code cid.Cid, initalizationParams interface{}) error
+// initInvocationContext is a more powerfull context for the "init" actor.
+type initInvocationContext interface {
+	InvocationContext
+	// Review: for @spec, added an error that is not yet on the spec but should be
+	// Note: @jzimmerman: "Agreed, this should support an error return."
+	CreateActor(cid cid.Cid, addr address.Address, constructorParams MethodParams) error
+}
 
+// minerInvocationContext has some special sauce for the miner.
+// Review: for @go-filecoin, do we want custom interfaces for the miners instead of grouping a bunch fo methods on the InvocationContext?
+type minerInvocationContext interface {
 	Verifier() verification.Verifier
 }
 
+// ActorStateHandler handles the actor state, allowing actors to lock on the state.
+type ActorStateHandler interface {
+	// Take returns the `Cid` for the state and locks the state.
+	//
+	// Any futre calls to `Take` by other instances will block until `Release` or `UpdateRelease` is called.
+	// Review: for @spec, the spec impl of `Take` is not doing any blocking, just panic'ing on the second call on the SAME object instance (this objects are not kept around, so it is always a new instance..).
```

Comment:
There's no concurrency, so blocking doesn't make sense. IMO this should cause the call to abort. Comparable to an assertion failure, indicating programmer error.

Reply:
Agreed.